text/0000-http-index.md

```diff
@@ -0,0 +1,132 @@
+- Feature Name: http_index
+- Start Date: 2019-10-18
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Selective download of the crates-io index over HTTP, similar to a solution used by Ruby's Bundler. Changes transport from an ahead-of-time Git clone to HTTP fetch as-needed, while keeping existing content and structure of the index. Most importantly, the proposed solution works with static files and doesn't require custom server-side APIs.
+
+# Motivation
+[motivation]: #motivation
+
+The full crate index is relatively big and slow to download. It will keep growing as crates.io grows, making the problem worse. The need to download the full index slows down the first use of Cargo. It's especially slow and wasteful in stateless CI environments, which download the full index, use only a tiny fraction of it, and throw it away. Caching of the index in hosted CI environments is difficult (`.cargo` dir is large) and often not effective (e.g. upload and download of large caches in Travis CI is almost as slow as a fresh index download).
+
+The kind of data stored in the index is not a good fit for the git protocol. The index content (as of eb037b4863) takes 176MiB as an uncompressed tarball, 16MiB with `gz -1`, and 10MiB compressed with `xz -6`. Git clone reports downloading 215MiB. That's more than just the uncompressed latest index content, and over **20 times more** than a compressed tarball.
+
+A while ago, GitHub indicated they [don't want to support shallow clones of large repositories](http://blog.cocoapods.org/Master-Spec-Repo-Rate-Limiting-Post-Mortem/). libgit2 doesn't support shallow clones yet. Squashing of the index history adds complexity to management and consumption of the index (which is also used by tools other than Cargo), and still doesn't solve problems of the git protocol inefficiency and overall growth.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Expose the index over HTTP as simple files, keeping the existing content and directory layout unchanged, e.g.:
```

Comment:
The index *is* currently exposed over HTTP as simple files, for example, https://raw.githubusercontent.com/rust-lang/crates.io-index/master/ac/ti/actions (without the cache directives we'd want of course)

Seems like a proof-of-concept cargo fork (that assumes files could be cached forever or using some other assumption) would be possible today, and would be a compelling addition to this RFC.

Reply:
Nice! The "raw" server does support Etag and gzip, so it may even be usable as-is!